woocommerce/woocommerce-ios
# Order Creation: Make UI immediately respond to additive changes to new order

Closes: #6339 
⚠️ Depends on #6949 ⚠️ 

## Description

During Order Creation, the new order screen will now immediately respond to additive changes (e.g. adding a new product to the order) by blocking the UI. Previously, there was a slight delay before the UI was blocked after making additive changes. This change will avoid scenarios where the user is able to open new screens to then be blocked while the network request completes.

## Changes

The previous delay happened because we called `.debounce()` first in `bindOrderSync()`, before updating the `state`. That allowed multiple changes to be made if they happened within the debounce period.

To make the new order screen immediately respond to additive changes, we move the `.debounce()` to later in the sync chain (in `bindOrderUpdate()` instead of `bindOrderSync()`). That way, we can update the `state` immediately as needed, before debouncing the changes to the order. This also allows us to increase the debounce time for non-blocking changes.

As a result:

* No debouncing is performed on the first sync, when the order is created remotely. This is always a blocking action.
* As additional change events are received, the state is immediately updated (in `.handleEvents()`) depending on whether the order contains new lines.
* After the state is updated, the events are debounced for 1 second (to limit the network requests made for non-blocking changes that may happen in quick succession, e.g. changing a product quantity).

### Test changes

The unit tests are updated to make sure the state publisher is subscribed to before making any changes to the order, since those state changes start happening immediately now. (Previously, it was ok to trigger the changes to the order first because they would be debounced, which allowed time for the subscription to the state publisher to be made and still receive all of the events.)

Also adds a unit test to check that double inputs are debounced when the order is being updated remotely. (I didn't add a test to check that the state updates immediately when there are additive changes to an order. We already have tests that check for that the correct state changes are made, and I couldn't think of a good approach to test that those changes happen immediately. Let me know if you have ideas for that!)

## Testing

1. Go to the Orders tab and create a new order.
2. Add a product, shipping, fee, or customer billing/shipping address to the order. Confirm the UI is immediately blocked (interactions are disabled and a spinner appears in the Payments section).
3. Make a non-additive change to the order, such as changing a product quantity or removing something from the order. Confirm the UI isn't blocked and you can make repeated changes, one right after the other.

## Screenshots

https://user-images.githubusercontent.com/8658164/170690579-d835685c-882d-440b-95ce-158c550bf560.mp4



## Submitter Checklist

Update release notes:
